### PR TITLE
Allow ComponentType to be passed to styled

### PR DESCRIPTION
**Pull request by pelotom**
This simplifies the typing of `styled` and allows passing it a `React.ComponentType<P>`. Adds a very simple test case that checks the same thing as #1420 but without the need to depend on all of `material-ui`.

Resolves #1420.
Resolves https://github.com/mui-org/material-ui/issues/9838.

### CODE REVIEW

**Comment by philpl** [comment]
cc @styled-components/typers: anyone available to take a look? seems like this would solve a lot of different reports we've been getting sporadically.

**Comment by NeoLegends** [comment]
Great work @pelotom.

I remember trying out the same (simplifying the interface with `React.ComponentType<T>`) but for some reason my changes weren't enough and I had issues with theme being optional. Happy this has been worked out, though!

File: typings/styled-components.d.ts [hunk 1]
```diff
@@ -53,9 +53,7 @@ type ThemedStyledComponentFactories<T> = {
 export interface ThemedBaseStyledInterface<T> extends ThemedStyledComponentFactories<T> {
   <P, TTag extends keyof JSX.IntrinsicElements>(tag: TTag): ThemedStyledFunction<P, T, P & JSX.IntrinsicElements[TTag]>;
   <P, O>(component: StyledComponentClass<P, T, O>): ThemedStyledFunction<P, T, O>;
-  <P extends { theme: T; }>(component: React.ComponentClass<P>): ThemedStyledFunction<P, T, WithOptionalTheme<P, T>>;
-  <P>(component: React.ComponentClass<P>): ThemedStyledFunction<P, T>;
-  <P extends { [prop: string]: any; theme?: T; }>(component: React.StatelessComponent<P>): ThemedStyledFunction<P, T, WithOptionalTheme<P, T>>;
+  <P extends { [prop: string]: any; theme?: T }>(component: React.ComponentType<P>): ThemedStyledFunction<P, T, WithOptionalTheme<P, T>>;
```
Comment (Igorbek) [inline typings/styled-components.d.ts]: Have you tried and tested that theme type `T` (type of `theme` prop) is actually respected by the compiler? I afraid having ` [prop: string]: any ` would eat all other props, including `theme`.

Comment (pelotom) [reply typings/styled-components.d.ts]: There are lots of existing tests for this; see [here](https://github.com/styled-components/styled-components/tree/e4f59b047cb92c693ed2218dca934eef703fa2e7/typings/tests/themed-tests). The `[prop: string]: any` trick (which you can see was already being used previously in one of the overloads) is necessary to ensure that a `theme` prop of the wrong type is not provided.

Comment (Igorbek) [reply typings/styled-components.d.ts]: Yes, makes sense. Just wanted to double-check.

**Comment by probablyup** [comment]
@Igorbek let's merge it?

**Comment by Igorbek** [comment]
@probablyup absolutely! I've no access to do it, go ahead and merge if you do.

**Comment by probablyup** [comment]
Thank you!

**Comment by mxstbr** [comment]
Thank you so much for helping us improve styled-components! Based on our [Community Guidelines](https://github.com/styled-components/styled-components/blob/master/CONTRIBUTING.md) every person that has a PR of any kind merged is offered **an invitation to the styled-components organization**—that is you right now!

Should you accept, you'll get write access to the main repository. (and a fancy `styled-components` logo on your GitHub profile!) You'll be able to label and close issues, merge pull requests, create new branches, etc. We want you to help us out with those things, so don't be scared to do them! Make sure to read our [contribution and community guidelines](https://github.com/styled-components/styled-components/blob/master/CONTRIBUTING.md), which explains all of this in more detail. Of course, if you have any questions just let me know!

**Pull request merged**